cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,229 @@
+
+    The logger implementation should be stateful, s.t.:
+        - Each call to `register_initial` registers a new transformer stage and initial circuit.
+        - Each subsequent call to `log` should store additional logs corresponding to the stage.
+        - Each call to `register_final` should register the end of the currently active stage.
+
+    The logger assumes that
+        - Transformers are run sequentially.
+        - Nested transformers are allowed, in which case the behavior would be similar to a
+          doing a depth first search on the graph of transformers -- i.e. the top level transformer
+          would end (i.e. receive a `register_final` call) once all nested transformers (i.e. all
+          `register_initial` calls received while the top level transformer was active) have
+          finished (i.e. corresponding `register_final` calls have also been received).
+        - This behavior can be simulated by maintaining a stack of currently active stages and
+          adding data from `log` calls to the stage at the top of the stack.
+
+    The `LogLevel`s can be used to control the input processing and output resolution of the logs.
+    """
+
+    @abc.abstractmethod
+    def register_initial(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        """Register the beginning of a new transformer stage.
+
+        Args:
+            circuit: Input circuit to the new transformer stage.
+            transformer_name: Name of the new transformer stage.
+        """
+
+    @abc.abstractmethod
+    def log(self, *args: str, level: LogLevel = LogLevel.INFO) -> None:
+        """Log additional metadata corresponding to the currently active transformer stage.
+
+        Args:
+            *args: The additional metadata to log.
+            level: Logging level to control the amount of metadata that gets put into the context.
+        """
+
+    @abc.abstractmethod
+    def register_final(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        """Register the end of the currently active transformer stage.
+
+        Args:
+            circuit: Final transformed output circuit from the transformer stage.
+            transformer_name: Name of the (currently active) transformer stage which ends.
+        """
+
+
+@dataclasses.dataclass()
+class TransformerContext:
+    """Stores common configurable options for transformers."""
+
+    logger: Optional[TransformerStatsLoggerBase] = None
+    ignore_tags: Tuple[Hashable, ...] = ()
+    max_parallelism: int = 1
+
+
+TRANSFORMER_TYPE = Callable[[CIRCUIT_TYPE, TransformerContext], CIRCUIT_TYPE]
+TRANSFORMER_TYPE_T = TypeVar(
+    'TRANSFORMER_TYPE_T',
+    TRANSFORMER_TYPE[circuit.Circuit],
+    TRANSFORMER_TYPE[frozen_circuit.FrozenCircuit],
+)
+TRANSFORMER_CLS_TYPE = TypeVar(
+    'TRANSFORMER_CLS_TYPE',
+    Type[TRANSFORMER_TYPE[frozen_circuit.FrozenCircuit]],
+    Type[TRANSFORMER_TYPE[circuit.Circuit]],
+)
+
+
+def _transformer_with_logging(
```

Comment:
nit: I'd suggest naming this with a verb phrase, since it actually _calls_ the transformer (name sounds like it produces a transformer with logging added). Perhaps `_transform_and_log`?

Reply:
done